# Nanodegree Engenheiro de Machine Learning
## Aprendizagem Supervisionada
## Project 2: Construindo um Sistema de Intervenção para Estudantes

Bem-vindo ao segundo projeto do Nanodegree de Machine Learning! Neste Notebook, alguns templates de código já foram fornecidos, e será o seu trabalho implementar funcionalidades necessárias para completar este projeto com êxito. Seções que começam com **'Implementação'** no cabeçalho indicam que o bloco de código que se segue precisará de funcionalidades adicionais que você deve fornecer. Instruções serão providenciadas para cada seção e as especificações para cada implementação estarão marcadas no bloco de código com o comando `'TODO'`. Tenha certeza de ler atentamente todas as instruções!

Além do código implementado, haverá questões relacionadas ao projeto e à implementação que você deve responder. Cada seção em que você tem que responder uma questão será antecedida de um cabeçalho **'Questão X'**. Leia atentamente cada questão e escreva respostas completas nas caixas de texto subsequentes que começam com **'Resposta: '**. O projeto enviado será avaliado baseado nas respostas para cada questão e a implementação que você forneceu.  

>**Nota:** Células de código e Markdown podem ser executadas utilizando o atalho de teclado **Shift + Enter**. Além disso, as células Markdown podem ser editadas, um clique duplo na célula entra no modo de edição.

### Questão 1 - Classificação versus Regressão
*Seu objetivo neste projeto é identificar estudantes que possam precisar de intervenção antecipada antes de serem reprovados. Que tipo de problema de aprendizagem supervisionada é esse: classificação ou regressão? Por quê?*

**Resposta: ** Classificação, em problemas de classificação estamos tentando encontrar uma saida discreta,ou seja, dada uma entrada o algoritimo mapeia a saida em uma categoria distinta. Para este problema em específico, estamos tentando prever quais estudantes serão aprovados ou reprovados, essas são as duas categorias possíveis.

## Observando os Dados
Execute a célula de código abaixo para carregar as bibliotecas de Python necessárias e os dados sobre os estudantes. Note que a última coluna desse conjunto de dados, `'passed'`, será nosso rótulo alvo (se o aluno foi ou não aprovado). As outras colunas são atributos sobre cada aluno.

In [1]:
# Importar bibliotecas
import numpy as np
import pandas as pd
from time import time
from sklearn.metrics import f1_score

# Ler os dados dos estudantes
student_data = pd.read_csv("student-data.csv")
print "Os dados dos estudantes foram lidos com êxito!"
student_data.head()


Os dados dos estudantes foram lidos com êxito!


school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   ...   internet romantic  famrel  freetime  goout Dalc Walc health absences  \
0  ...         no       no       4         3      4    1    1      3        6   
1  ...        yes       no       5         3      3    1    1      3        4   
2  ...        yes       no       4         3      2    2    3      3       10   
3  ...        yes      yes       3         2      2    1    1      5        2   
4  ...         no       no       4         3      2    1    2      5        4   

  passed  
0     no  
1     no  
2    yes  
3    yes  
4    yes  

[5 rows x 31 columns]

### Implementação: Observando os Dados
Vamos começar observando o conjunto de dados para determinar quantos são os estudantes sobre os quais temos informações e entender a taxa de graduação entre esses estudantes. Na célula de código abaixo, você vai precisar calcular o seguinte:
- O número total de estudantes, `n_students`.
- O número total de atributos para cada estudante, `n_features`.
- O número de estudantes aprovados, `n_passed`.
- O número de estudantes reprovados, `n_failed`.
- A taxa de graduação da classe, `grad_rate`, em porcentagem (%).


In [2]:
# TODO: Calcule o número de estudante
n_students = len(student_data.index)

# TODO: Calcule o número de atributos
n_features = len(student_data.columns) -1 # because the column passed is not a feature

# TODO: Calcule o número de alunos aprovados
n_passed = student_data[student_data['passed'] == 'yes'].shape[0]

# TODO: Calcule o número de alunos reprovados
n_failed = student_data[student_data['passed'] == 'no'].shape[0]

# TODO: Calcule a taxa de graduação
grad_rate = float(n_passed*1.0 / n_students*1.0)*100

# Imprima os resultados
print "Número total de estudantes: {}".format(n_students)
print "Número de atributos: {}".format(n_features)
print "Número de estudantes aprovados: {}".format(n_passed)
print "Número de estudantes reprovados: {}".format(n_failed)
print "Taxa de graduação: {:.2f}%".format(grad_rate)

Número total de estudantes: 395
Número de atributos: 30
Número de estudantes aprovados: 265
Número de estudantes reprovados: 130
Taxa de graduação: 67.09%


## Preparando os Dados
Nesta seção, vamos preparara os dados para modelagem, treinamento e teste.

### Identificar atributos e variáveis-alvo
É comum que os dados que você obteve contenham atributos não numéricos. Isso pode ser um problema, dado que a maioria dos algoritmos de machine learning esperam dados númericos para operar cálculos.

Execute a célula de código abaixo para separar os dados dos estudantes em atributos e variáveis-alvo e verificar se algum desses atributos é não numérico.

In [3]:
# Extraia as colunas dos atributo
feature_cols = list(student_data.columns[:-1])

# Extraia a coluna-alvo, 'passed'
target_col = student_data.columns[-1] 

# Mostre a lista de colunas
print "Colunas de atributos:\n{}".format(feature_cols)
print "\nColuna-alvo: {}".format(target_col)

# Separe os dados em atributos e variáveis-alvo (X_all e y_all, respectivamente)
X_all = student_data[feature_cols]
y_all = student_data[target_col]

# Mostre os atributos imprimindo as cinco primeiras linhas
print "\nFeature values:"
print X_all.head()

# print do y_all
print "\nTarget:"
print y_all.head()

Colunas de atributos:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']

Coluna-alvo: passed

Feature values:
  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

    ...    higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0   ...       yes       no        no       4         3     4    1    1      3   
1   ...    

### Pré-processar Colunas de Atributo

Como você pode ver, há muitas colunas não numéricas que precisam ser convertidas! Muitas delas são simplesmente `yes`/`no`, por exemplo, a coluna `internet`. É razoável converter essas variáveis em valores (binários) `1`/`0`.

Outras colunas, como `Mjob` e `Fjob`, têm mais do que dois valores e são conhecidas como variáveis categóricas. A maneira recomendada de lidar com esse tipo de coluna é criar uma quantidade de colunas proporcional aos possíveis valores (por exemplo, `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc), e assinalar `1` para um deles e `0` para todos os outros.

Essas colunas geradas são por vezes chamadas de _variáveis postiças_ (em inglês: _dummy variables_), e nós iremos utilizar a função [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) para fazer essa conversão. Execute a célula de código abaixo para executar a rotina de pré-processamento discutida nesta seção.

In [4]:
def preprocess_features(X):
    ''' Pré-processa os dados dos estudantes e converte as variáveis binárias não numéricas em
        variáveis binárias (0/1). Converte variáveis categóricas em variáveis postiças. '''
    
    # Inicialize nova saída DataFrame
    output = pd.DataFrame(index = X.index)

    # Observe os dados em cada coluna de atributos 
    for col, col_data in X.iteritems():
        
        # Se o tipo de dado for não numérico, substitua todos os valores yes/no por 1/0
        if col_data.dtype == object:
            col_data = col_data.replace(['yes', 'no'], [1, 0])

        # Se o tipo de dado for categórico, converta-o para uma variável dummy
        if col_data.dtype == object:
            # Example: 'school' => 'school_GP' and 'school_MS'
            col_data = pd.get_dummies(col_data, prefix = col)  
        
        # Reúna as colunas revisadas
        output = output.join(col_data)
    
    return output

X_all = preprocess_features(X_all)
print "Processed feature columns ({} total features):\n{}".format(len(X_all.columns), list(X_all.columns))

Processed feature columns (48 total features):
['school_GP', 'school_MS', 'sex_F', 'sex_M', 'age', 'address_R', 'address_U', 'famsize_GT3', 'famsize_LE3', 'Pstatus_A', 'Pstatus_T', 'Medu', 'Fedu', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services', 'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other', 'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home', 'reason_other', 'reason_reputation', 'guardian_father', 'guardian_mother', 'guardian_other', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']


### Implementação: Divisão dos Dados de Treinamento e Teste
Até agora, nós convertemos todos os atributos _categóricos_ em valores numéricos. Para o próximo passo, vamos dividir os dados (tanto atributos como os rótulos correspondentes) em conjuntos de treinamento e teste. Na célula de código abaixo, você irá precisar implementar o seguinte:
- Embaralhe aleatoriamente os dados (`X_all`, `y_all`) em subconjuntos de treinamento e teste.
  - Utilizar 300 pontos de treinamento (aproxidamente 75%) e 95 pontos de teste (aproximadamente 25%).
  - Estabelecer um `random_state` para as funções que você utiliza, se a opção existir.
  - Armazene os resultados em `X_train`, `X_test`, `y_train` e `y_test`.

In [5]:
# TODO: Importe qualquer funcionalidade adicional de que você possa precisar aqui
from sklearn.model_selection import train_test_split

# TODO: Estabeleça o número de pontos de treinamento
num_train = 300

# Estabeleça o número de pontos de teste
num_test = X_all.shape[0] - num_train

# TODO: Emabaralhe e distribua o conjunto de dados de acordo com o número de pontos de treinamento e teste abaixo
X_train, X_test, y_train, y_test = train_test_split(X_all, y_all, test_size=num_test, train_size=num_train, random_state=9)

# Mostre o resultado da distribuição
print "O conjunto de treinamento tem {} amostras.".format(X_train.shape[0])
print "O conjunto de teste tem {} amostras.".format(X_test.shape[0])

O conjunto de treinamento tem 300 amostras.
O conjunto de teste tem 95 amostras.


## Treinando e Avaliando Modelos
Nesta seção, você irá escolher 3 modelos de aprendizagem supervisionada que sejam apropriados para esse problema e que estejam disponíveis no `scikit-learn`. Primeiro você irá discutir o raciocínio por trás da escolha desses três modelos considerando suas vantagens e desvantagens e o que você sabe sobre os dados. Depois você irá ajustar o modelo a diferentes tamanhos de conjuntos de treinamento (com 100, 200 e 300 pontos) e medir a pontuação F<sub>1</sub>. Você vai precisar preencher três tabelas (uma para cada modelo) que mostrem o tamanho do conjunto de treinamento, o tempo de treinamento, o tempo de previsão e a pontuação F<sub>1</sub> no conjunto de treinamento.

**Os seguintes modelos de aprendizagem supervisionada estão atualmente disponíveis no **[`scikit-learn`](http://scikit-learn.org/stable/supervised_learning.html)** para você escolher:**
- Gaussian Naive Bayes (GaussianNB)
- Árvores de Decisão
- Métodos de agregação (Bagging, AdaBoost, Random Forest, Gradient Boosting)
- K-Nearest Neighbors (KNeighbors)
- Método do gradiente estocástico (SGDC)
- Máquinas de vetores de suporte (SVM)
- Regressão logística

### Questão 2 - Aplicação dos Modelos
*Liste três modelos de aprendizagem supervisionada que são apropriadas para esse problema. Para cada modelo escolhido:*
- Descreva uma aplicação em mundo real na indústria em que o modelo pode ser aplicado. *(Talvez você precise fazer um pouco de pesquisa para responder essa questão – dê as devidas referências!)* 
- Quais são as vantagens do modelo; quando ele tem desempenho melhor? 
- Quais são as desvantagens do modelo, quando ele tem desempenho pior?
- O que faz desse modelo um bom candidato para o problema, considerando o que você sabe sobre os dados?

**Resposta: ** Os três modelos escolhidos são: 1 - Maquina de vetores de suporte (SVM), 2 - Random Forests, 3 - Gaussian Naive Bayes

### 1 - Maquina de vetores de suporte

O SVM padrão toma como entrada um conjunto de dados e prediz, para cada entrada dada, qual de duas possíveis classes a entrada faz parte, o que faz do SVM um classificador linear binário não probabilístico. 

#### **Aplicação:**
    
SVM utilizado para aplicação na avaliação da solvencia de cooperativas de créditos.
    
__fonte__: (https://repositorio.ufsc.br/xmlui/bitstream/handle/123456789/158361/318899.pdf?sequence=1&isAllowed=y)

##### **Vantagens**
>- Em caso de outlier a SVM busca a melhor forma possível de classificação e, se necessário, desconsidera o outlier;
- Funciona muito bem em domínios complicados, em que existe uma clara margem de separação;
- Boa capacidade de generalização: os classificadores gerados
por uma SVM geralmente alcançam bons resultados de generalização,
essa capacidade é medida por sua eficiência na classificação de dados
que não pertençam ao conjunto utilizado no seu treino.
- Robustez em grandes dimensões: as SVMs são robustas diante
de objetos de grandes dimensões.
- Convexidade da função objetiva: a aplicação das SVMs implica
a otimização de uma função quadrática que possui apenas um mínimo
global. Essa é uma vantagem sobre, por exemplo, as Redes Neurais
Artificiais, em que há presença de mínimos locais na função objetiva a
ser minimizada.

##### **Desvantagens**
>- Não funciona bem em conjuntos de dados muito grandes, pois exige inversão de matriz - aumentando a complexidade computacional com até o cubo do volume de dados;
- Não funciona bem em conjunto de dados com grande quantidade de ruídos;

A escolha do algoritmo SVM é adequado para este aplicativo porque os dados contém muitos recursos. Usar o kernel para tornar esses recursos linearmente separáveis será efetivo para classificar estudantes passando/não passando.

### 2 - Random Forests
O algoritmo Random Forest é um tipo de ensemble learning,
utliza um método que gera muitos classificadores e combina o seus resultados.

Nesse caso ele gera vários decision trees, cada um com suas particularidades, e combina o resultado da classificação de todos eles. 
Essa combinação de modelos, orna ele um algoritmo muito mais poderoso do que o Decision Tree.

#### **Aplicação:**

A avaliação do método baseado em RF foi demonstrada por um estudo de caso sobre o diagnóstico de falhas dos motores de indução.
Dentro o motivo da escolha do algotirimo, podemos citar:
a velocidade executada rapidamente, a característica do classificador de árvores e o alto desempenho no diagnóstico de falhas da máquina.

__fonte__: (https://link.springer.com/chapter/10.1007/978-1-84628-814-2_82)

##### **Vantagens**

>- Alto nível de acurácia;
- Simplicidade de construção, e fácil entendimento.
- Pode ter entradas discretas ou continuas usando usando intervalos como limites de decisão.
- A natureza aleatoria de construçãao de cada árvore minimiza o sobreajuste;
- Tende a ter bom desempenho com dados desbalanceados.
- Seleciona um subconjunto aleatório de recursos em cada divisão. A idéia é que os recursos fortes serão naturalmente selecionados por múltiplas árvores, fazendo com que elas se tornem dominantes quando as árvores são mescladas.

##### **Desvantagens**

>- Construção e manipulação de várias árvores.
- Uma das limitações mais importantes das DT é relacionado ao overfitting devido a natureza do 
classificador, onde a fronteira de decisão tende a ser paralelo aos eixos existentes do plano de amostras. Sem um ajuste de pruning, por exemplo, podem haver distorções na fronteira de decisão que pode prejudicar distribuições suaves.
- Possível crescimento do nó exponencial à medida que mais recursos são adicionados (n!).
- Possível crescimento exponencial da folha à medida que mais nódulos são gerados (n2).
- Múltiplas árvores na floresta aleatória aumentarão os recursos computacionais e o tempo para formação e previsão.

Random Forest é um algoritmo geral a ser considerado ao lidar com o problema da classificação, também neste caso, devemos lidar com os 
principais recursos dimensionais.

### 3 - Gaussian Naive Bayes

Em aprendizagem de maquinas algoritimos de classificação Naive Bayes são uma família de simples classificadores probabilísticos com base na aplicação do teorema de Bayes (naive).

Um dos motivos ao escolher esse algoritimo, alem da simplicidade de implementação e entendimento, é a quantiodade de exemplos disponiveis que contribui bastante para ter um ponto de partido.

#### **Aplicação:**
Nesse exemplo é utilizado a classificação de Naive Bayes para reconhecer emoções através de expressões faciais.

__fonte__: (https://pdfs.semanticscholar.org/7816/e94ff9e07fcde3e09c7e55ea8d2357eb69ec.pdf)

##### **Vantagens**
>- Funciona especialmente bem em ambientes de baixa polarização/baixa variação
- Eficaz em conjuntos de dados pequenos
-Este algoritmo tem como principais vantagens a independência das variáveis envolvidas, ou seja, cada característica é capaz de contribuir individulamente para a probabilidade de cálculo de determinadas classes.

##### **Desvantagens**
>-Possui a desvatangem de assumir independência de variáveis quando no mundo real elas tendem a ser dependentes.
- Menos eficaz à medida que os dados se tornam maiores e mais complexos

O conjunto de treinamento neste exemplo é relativamente pequeno e a principio muitos dos recursos parecem ser independentes, esse é o principal motivo por testar essa implementação.

### Configuração
Execute a célula de código abaixo para inicializar três funções de ajuda que você pode utilizar para treinar e testar os três modelos de aprendizagem supervisionada que você escolheu acima. As funções são as seguintes:
- `train_classifier` - recebe como parâmetro um classificador e dados de treinamento e ajusta o classificador aos dados.
- `predict_labels` - recebe como parâmetro um classificador ajustado, atributos e rótulo alvo e faz estimativas utilizando a pontuação do F<sub>1</sub>.
- `train_predict` - recebe como entrada um classificador, e dados de treinamento e teste, e executa `train_clasifier` e `predict_labels`.
 - Essa função vai dar a pontuação F<sub>1</sub> tanto para os dados de treinamento como para os de teste, separadamente.

In [6]:
def train_classifier(clf, X_train, y_train):
    ''' Ajusta um classificador para os dados de treinamento. '''
    
    # Inicia o relógio, treina o classificador e, então, para o relógio
    start = time()
    clf.fit(X_train, y_train)
    end = time()
    
    # Imprime os resultados
    print "O modelo foi treinado em {:.4f} segundos".format(end - start)

    
def predict_labels(clf, features, target):
    ''' Faz uma estimativa utilizando um classificador ajustado baseado na pontuação F1. '''
    
    # Inicia o relógio, faz estimativas e, então, o relógio para
    start = time()
    y_pred = clf.predict(features)
    end = time()
    
    # Imprime os resultados de retorno
    print "As previsões foram feitas em {:.4f} segundos.".format(end - start)
    return f1_score(target.values, y_pred, pos_label='yes')


def train_predict(clf, X_train, y_train, X_test, y_test):
    ''' Treina e faz estimativas utilizando um classificador baseado na pontuação do F1. '''
    
    # Indica o tamanho do classificador e do conjunto de treinamento
    print "Treinando um {} com {} pontos de treinamento. . .".format(clf.__class__.__name__, len(X_train))
    
    # Treina o classificador
    train_classifier(clf, X_train, y_train)
    
    # Imprime os resultados das estimativas de ambos treinamento e teste
    print "Pontuação F1 para o conjunto de treino: {:.4f}.".format(predict_labels(clf, X_train, y_train))
    print "Pontuação F1 para o conjunto de teste: {:.4f}.".format(predict_labels(clf, X_test, y_test))

### Implementação: Métricas de Desempenho do Modelo
Com as funções acima, você vai importar os três modelos de aprendizagem supervisionada de sua escolha e executar a função `train_prediction` para cada um deles. Lembre-se de que você vai precisar treinar e usar cada classificador para três diferentes tamanhos de conjuntos de treinamentos: 100, 200 e 300 pontos. Então você deve ter 9 saídas diferentes abaixo – 3 para cada modelo utilizando cada tamanho de conjunto de treinamento. Na célula de código a seguir, você deve implementar o seguinte:
- Importe os três modelos de aprendizagem supervisionada que você escolheu na seção anterior.
- Inicialize os três modelos e armazene eles em `clf_A`, `clf_B` e `clf_C`.
 - Defina um `random_state` para cada modelo, se a opção existir.
 - **Nota:** Utilize as configurações padrão para cada modelo – você vai calibrar um modelo específico em uma seção posterior.
- Crie diferentes tamanhos de conjuntos de treinamento para treinar cada modelo.
 - *Não embaralhe e distribua novamente os dados! Os novos pontos de treinamento devem ser tirados de `X_train` e `y_train`.*
- Treine cada modelo com cada tamanho de conjunto de treinamento e faça estimativas com o conjunto de teste (9 vezes no total).  
**Nota:** Três tabelas são fornecidas depois da célula de código a seguir, nas quais você deve anotar seus resultados.

In [7]:
from sklearn import svm
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB

clf_A = svm.SVC(random_state=0)
clf_B = RandomForestClassifier(random_state=0)
clf_C = GaussianNB()

# TODO: Configure os tamanho dos conjuntos de treinamento
X_train_100 = X_train[:100]
y_train_100 = y_train[:100]

X_train_200 = X_train[:200]
y_train_200 = y_train[:200]

X_train_300 = X_train
y_train_300 = y_train

train_predict(clf_A, X_train_100, y_train_100, X_test, y_test)
print
train_predict(clf_A, X_train_200, y_train_200, X_test, y_test)
print
train_predict(clf_A, X_train_300, y_train_300, X_test, y_test)
print
print "---------------"
print
train_predict(clf_B, X_train_100, y_train_100, X_test, y_test)
print
train_predict(clf_B, X_train_200, y_train_200, X_test, y_test)
print
train_predict(clf_B, X_train_300, y_train_300, X_test, y_test)
print
print "---------------"
print
train_predict(clf_C, X_train_100, y_train_100, X_test, y_test)
print
train_predict(clf_C, X_train_200, y_train_200, X_test, y_test)
print
train_predict(clf_C, X_train_300, y_train_300, X_test, y_test)

Treinando um SVC com 100 pontos de treinamento. . .
O modelo foi treinado em 0.0020 segundos
As previsões foram feitas em 0.0011 segundos.
Pontuação F1 para o conjunto de treino: 0.9041.
As previsões foram feitas em 0.0010 segundos.
Pontuação F1 para o conjunto de teste: 0.8375.

Treinando um SVC com 200 pontos de treinamento. . .
O modelo foi treinado em 0.0043 segundos
As previsões foram feitas em 0.0033 segundos.
Pontuação F1 para o conjunto de treino: 0.8669.
As previsões foram feitas em 0.0016 segundos.
Pontuação F1 para o conjunto de teste: 0.8442.

Treinando um SVC com 300 pontos de treinamento. . .
O modelo foi treinado em 0.0080 segundos
As previsões foram feitas em 0.0071 segundos.
Pontuação F1 para o conjunto de treino: 0.8472.
As previsões foram feitas em 0.0028 segundos.
Pontuação F1 para o conjunto de teste: 0.8387.

---------------

Treinando um RandomForestClassifier com 100 pontos de treinamento. . .
O modelo foi treinado em 0.0265 segundos
As previsões foram feitas em

### Resultados Tabulados
Edite a célula abaixo e veja como a tabela pode ser desenhada em [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables). Você deve salvar seus resultados abaixo nas tabelas fornecidas.

** Classificador 1 - SVM**  

| Tamanho do Conjunto de Treinamento | Tempo de Treinamento | Tempo de Estimativa (teste) | Pontuação F1 (treinamento) | Pontuação F1 (teste) |
| :--------------------------------: | :------------------: | :-------------------------: | :------------------------: | :------------------: |
| 100                                |      0.0020          |        0.0014               |        0.9041              |        0.8375        |
| 200                                |      0.0052          |        0.0040               |        0.8669              |        0.8442        |
| 300                                |      0.0095          |        0.0066               |        0.8472              |        0.8387        |

** Classificador 2 - Random Forest**  

| Tamanho do Conjunto de Treinamento | Tempo de Treinamento | Tempo de Estimativa (teste) | Pontuação F1 (treinamento) | Pontuação F1 (teste) |
| :--------------------------------: | :------------------: | :-------------------------: | :------------------------: | :------------------: |
| 100                                |       0.0315         |            0.0016           |           1.0000           |        0.8028        |
| 200                                |       0.0234         |            0.0012           |           0.9760           |        0.8175        |
| 300                                |       0.0204         |            0.0013           |           0.9949           |        0.7681        |

** Classificador 3 - GaussianNB**  

| Tamanho do Conjunto de Treinamento | Tempo de Treinamento | Tempo de Estimativa (teste) | Pontuação F1 (treinamento) | Pontuação F1 (teste) |
| :--------------------------------: | :------------------: | :-------------------------: | :------------------------: | :------------------: |
| 100                                |       0.0009         |          0.0003             |          0.2857            |        0.2588       |
| 200                                |       0.0008         |          0.0003             |          0.7559            |        0.7910        |
| 300                                |       0.0014         |          0.0005             |          0.7931            |        0.8429        |

## Escolhendo o Melhor Modelo
Nesta seção final, você irá escolher dos três modelos de aprendizagem supervisionada o *melhor* para utilizar os dados dos estudantes. Você então executará um busca em matriz otimizada para o modelo em todo o conjunto de treinamento (`X_train` e `y_train`) ao calibrar pelo menos um parâmetro, melhorando em comparação a pontuação F<sub>1</sub> do modelo não calibrado. 

### Questão 3 - Escolhendo o Melhor Modelo
*Baseando-se nos experimentos que você executou até agora, explique em um ou dois parágrafos ao conselho de supervisores qual modelo que você escolheu como o melhor. Qual modelo é o mais apropriado baseado nos dados disponíveis, recursos limitados, custo e desempenho?*

**Resposta: **

Com base nos experimentos realizados acima, o algoritmo da Máquina de Vetores de Suporte (SVM) é mais apropriado.

À medida que os dados de treinamento aumentam de tamanho, ele cria um modelo mais generalizado com um crescente escore F1 nos dados de teste. 

Ele tem o desempenho computacional mais lento, mas esse custo é aceitável dada a capacidade superior de generalizar os dados.


### Questão 4 – O Modelo para um Leigo
*Em um ou dois parágrafos, explique para o conselho de supervisores, utilizando termos leigos, como o modelo final escolhido deve trabalhar. Tenha certeza que você esteja descrevendo as melhores qualidades do modelo, por exemplo, como o modelo é treinado e como ele faz uma estimativa. Evite jargões técnicos ou matemáticos, como descrever equações ou discutir a implementação do algoritmo.*

**Resposta: ** 

O suporte das máquinas vetoriais funciona focando em dados que são mais difíceis de classificar, como dois pontos próximos a um limite de decisão.

![alt text](svm0.png "Title")

Se esses dois pontos podem ser classificados com precisão, a intuição nos dirá que a mesma fórmula pode ser aplicada ao restante dos dados. Em resumo, um SVM tenta dar sentido a dados que são difíceis de distinguir, e então usa essa lógica para fazer previsões entre todo o conjunto de dados.

Essa previsão é feita com técnica conhecida como o truque do kernel, que transforma a forma dos dados para se tornar mais linear.

Começa por identificar os dois pontos mais próximos, conhecidos como vetores de suporte e, em seguida, tomar a diferença entre eles. 

![alt text](svm.png "Title")

A distância entre cada ponto de dados é maximizada, enquanto a linha de separação ideal será executada diretamente 
através desses dois pontos. Esta linha maximiza a margem entre cada ponto. 





### Implementação: Calibrando o Modelo
Calibre o modelo escolhido. Utilize busca em matriz (`GridSearchCV`) com, pelo menos, um parâmetro importante calibrado com, pelo menos, 3 valores diferentes. Você vai precisar utilizar todo o conjunto de treinamento para isso. Na célula de código abaixo, você deve implementar o seguinte:
- Importe [`sklearn.grid_search.gridSearchCV`](http://scikit-learn.org/stable/modules/generated/sklearn.grid_search.GridSearchCV.html) e [`sklearn.metrics.make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html).
- Crie o dicionário de parâmetros que você deseja calibrar para o modelo escolhido.
 - Examplo: `parameters = {'parameter' : [list of values]}`.
- Inicialize o classificador que você escolheu e armazene-o em `clf`.
- Crie a função de pontuação F<sub>1</sub> utilizando `make_scorer` e armazene-o em `f1_scorer`.
 - Estabeleça o parâmetro `pos_label` para o valor correto!
- Execute uma busca em matriz no classificador `clf` utilizando o `f1_scorer` como método de pontuação e armazene-o em `grid_obj`.
- Treine o objeto de busca em matriz com os dados de treinamento (`X_train`, `y_train`) e armazene-o em `grid_obj`.

In [8]:
#TODO: Importe 'GridSearchCV' e 'make_scorer'
from sklearn import grid_search
from sklearn import metrics

# TODO: Crie a lista de parâmetros que você gostaria de calibrar
parameters = {'kernel':('rbf', 'linear'), 'C': (0.5, 1.0, 2.0, 5.0, 10.0), 'gamma' : ( 0.001, 1, 10) }

# TODO: Inicialize o classificador
clf = svm.SVC()

# TODO: Faça uma função de pontuação f1 utilizando 'make_scorer' 
f1_scorer = metrics.make_scorer(metrics.f1_score, pos_label='yes')

# TODO: Execute uma busca em matriz no classificador utilizando o f1_scorer como método de pontuação
grid_obj = grid_search.GridSearchCV(clf, parameters, scoring=f1_scorer)

# TODO: Ajuste o objeto de busca em matriz para o treinamento de dados e encontre os parâmetros ótimos
grid_obj = grid_obj.fit(X_train, y_train.values.ravel())

# Get the estimator
clf = grid_obj.best_estimator_

print "Melhores"
print grid_obj.best_params_

# Reporte a pontuação final F1 para treinamento e teste depois de calibrar os parâmetrosprint "Tuned model has a training F1 score of {:.4f}.".format(predict_labels(clf, X_train, y_train))
print "O modelo calibrado tem F1 de {:.4f} no conjunto de treinamento.".format(predict_labels(clf, X_train, y_train))
print "O modelo calibrado tem F1 de {:.4f} no conjunto de teste.".format(predict_labels(clf, X_test, y_test))

/home/gustavo/anaconda2/lib/python2.7/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)
/home/gustavo/anaconda2/lib/python2.7/site-packages/sklearn/grid_search.py:42: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. This module will be removed in 0.20.
  DeprecationWarning)


Melhores
{'kernel': 'rbf', 'C': 10.0, 'gamma': 0.001}
As previsões foram feitas em 0.0060 segundos.
O modelo calibrado tem F1 de 0.8205 no conjunto de treinamento.
As previsões foram feitas em 0.0020 segundos.
O modelo calibrado tem F1 de 0.8645 no conjunto de teste.


### Questão 5 - Pontuação F<sub>1</sub> Final
*Qual é a pontuação F<sub>1</sub> do modelo final para treinamento e teste? Como ele se compara ao modelo que não foi calibrado?*

**Resposta: **

A pontuação F1 do modelo final para treinamento é 0.8205, e para testes é 0.8645. A pontuação melhorou com relação ao modelo calibrado em teste.


> **Nota**: Uma vez que você completou todas as implementações de código e respondeu todas as questões acima com êxito, você pode finalizar seu trabalho exportando o iPython Nothebook como um document HTML. Você pode fazer isso utilizando o menu acima e navegando para  
**File -> Download as -> HTML (.html)**. Inclua a documentação final junto com o notebook para o envio do seu projeto.